linkerd/core/src/main/scala/io/buoyant/linkerd/RequestActiveTracer.scala

```diff
@@ -0,0 +1,198 @@
+  namer: DstBindingFactory.Namer,
+  dtab: BaseDtab,
+  routerLabel: String
+) extends SimpleFilter[Request, Response] {
+
+  private[this] case class Node(path: String, dentry: String, treeNode: DelegateTree[_]) {
+    override def toString: String = {
+      treeNode match {
+        case DelegateTree.Delegate(_, d, _) if d == Dentry.nop => s"$path"
+        case _ => s"$path ($dentry)"
+      }
+
+    }
+  }
+
+  private[this] val RequestTracerMaxDepthHeader = "l5d-max-depth"
+
+  private[this] def printEvaluatedRequest(
+    prependText: String,
+    serviceName: String,
+    clientName: String,
+    selectedAddress: String,
+    addresses: Option[Set[String]],
+    dtabResolution: List[String]
+  ): String = {
+    Seq(
+      prependText,
+      s"""
+         |--- Router: $routerLabel ---
+         |service name: $serviceName
+         |client name: $clientName
+         |addresses: [${addresses.getOrElse(Set.empty).mkString(", ")}]
+         |selected address: $selectedAddress
+         |dtab resolution:
+         |${dtabResolution.map("  " + _).mkString("\n")}
+    """.stripMargin
+    ).mkString("\n").trim.concat("\n")
+  }
+
+  /**
+   * formatDelegation prints out path the proxy used to resolve the client name.
+   * It 'walks' a DelegateTree of Name.Bound and finds a DelegateLeaf that matches
+   * the client path provided.
+   * @param dTree
+   * @param tree
+   * @param searchPath
+   * @return
+   */
+  private[this] def formatDelegation(
+    dTree: DelegateTree[Name.Bound],
+    tree: List[Node],
+    searchPath: Path
+  ): List[Node] = {
+    dTree match {
+      case l@Leaf(leafPath, dentry, bound) if bound.id == searchPath =>
+        tree.lastOption.map {
+          case Node(nodePath, nodeDentry, Transformation(_, _, _, _)) =>
+            tree.dropRight(1) :+ Node(
+              nodePath,
+              dentry.show,
+              DelegateTree.Empty(Path.empty, Dentry.nop)
+            ) :+ Node(leafPath.show, nodeDentry, DelegateTree.Empty(Path.empty, Dentry.nop))
+          case _ => tree :+ Node(leafPath.show, dentry.show, l)
+        }.getOrElse(List.empty)
+      case t@Transformation(path, name, _, remainingTree) =>
+        formatDelegation(remainingTree, tree :+ Node(path.show, name, t), searchPath)
+      case d@Delegate(path, dentry, remainingTree) =>
+        formatDelegation(remainingTree, tree :+ Node(path.show, dentry.show, d), searchPath)
+      case u@Union(path, dentry, remainingWeightedTrees@_*) =>
+        remainingWeightedTrees.map { wd =>
+          formatDelegation(wd.tree, tree :+ Node(path.show, dentry.show, u), searchPath)
+        }.find(!_.isEmpty).toList.flatten
+      case a@Alt(path, dentry, remainingTrees@_*) =>
+        remainingTrees.map { d =>
+          formatDelegation(d, tree :+ Node(path.show, dentry.show, a), searchPath)
+        }.find(!_.isEmpty).toList.flatten
+      case _ => List.empty
+    }
+  }
+
+  private[this] def getRequestTraceResponse(prevResp: Response) = {
```

Comment:
might just call this `response` instead of `prevResp` since we're mutating it and not creating a new one